Feat/floatingip

ezeriver94 opened the pull request:
Implement FloatingIP controller/actuator

mandre commented on cmd/resource-generator/main.go:
```diff
@@ -61,6 +61,17 @@ var allResources []templateFields = []templateFields{
 	},
+	{
+		Name: "FloatingIP",
+		AdditionalPrintColumns: []additionalPrintColumn{
```
Nice!

mandre reviewed (commented):
Thanks for the PR! This is great.

Early feedback on the API for now. I haven't looked at the rest yet.

mandre commented:
> Hi @mandre, thanks for the quick feedback. I was going to address the lint issues before I ask for review but I didn't have much time to look at it. I'll go over your comments and lint validations these days. Regarding 'Label PR / semver' job, I'm not sure how to fix it, if you could give me some hints I would appreciate it.

It could not rebase your PR due to a conflict: https://github.com/k-orc/openstack-resource-controller/actions/runs/14848877163/job/41688586737?pr=392#step:3:11

You should rebase manually on top of main and fix the conflict.

ezeriver94 commented:
> You should do a [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) (and not a merge) because go-apidiff rebases your changes and is not able to apply one of them, you need to update it so it applies cleanly on top of main.

Thanks, looks better now.
FYI I've just pushed a change into [adapter.template](https://github.com/ezeriver94/openstack-resource-controller/blob/feat/floatingip/cmd/resource-generator/data/adapter.template#L79) to support non-named Openstack resources. Before I added Name into FloatingIP ResourceSpec but as there is no name in Openstack side, it was not really a good solution. Let me know if this does not make sense for you

mandre reviewed (commented):
Still not finished with the review, but I wanted to leave my pending comments before my extended weekend :smiley:

ezeriver94 commented on internal/controllers/floatingip/actuator.go:
```diff
@@ -0,0 +1,216 @@
+}
+
+func (actuator floatingipActuator) ListOSResourcesForAdoption(ctx context.Context, obj *orcv1alpha1.FloatingIP) (floatingipIterator, bool) {
+	// We don't support adoption of floatingips as they don't have a name
```
Definitely, at the beginning I hadn't added FloatingIP as resource filter, and then I forgot to support it during adoption.

ezeriver94 commented on internal/controllers/floatingip/tests/floatingip-create-full/00-prerequisites.yaml:
```diff
@@ -0,0 +1,8 @@
+commands:
+  - command: kubectl create secret generic openstack-clouds --from-file=clouds.yaml=${E2E_KUTTL_OSCLOUDS} ${E2E_KUTTL_CACERT_OPT}
+    namespaced: true
+  - script: cat ../templates/import-external-network.yaml |envsubst | kubectl -n ${NAMESPACE} apply -f -
```
ok, I add it as I didn't have the possibility to create external networks in the project I was testing with. I can add the tests to create the external networks as well (removing templates) and look at the outcome in this PR's actions

ezeriver94 commented on api/v1alpha1/floatingip_types.go:
```diff
@@ -0,0 +1,109 @@
+	// fixedIP is the IP address of the port to which the floatingip is associated.
+	// +optional
+	FixedIP *IPvAny `json:"fixedIP,omitempty"`
+}
```
Ok, maybe for a next PR 😅

ezeriver94 commented on api/v1alpha1/floatingip_types.go:
```diff
@@ -0,0 +1,109 @@
+
+// FloatingIPResourceSpec contains the desired state of a floating IP
+// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="FloatingIPResourceSpec is immutable"
+type FloatingIPResourceSpec struct {
```
I would like too but I don't have much time to spend on this 😞  also, this is my first experience contributing to an open source project (as you may have noticed) so it would probably take some time to get it merged

ezeriver94 commented on internal/controllers/floatingip/tests/templates/import-external-floating-ip.yaml:
```diff
@@ -0,0 +1,13 @@
+  managementPolicy: unmanaged
+  import:
+    filter:
+      floatingIP: "${E2E_EXTERNAL_FLOATING_IP_ADDRESS}"
```
removed in favor of creating dependencies without templates

ezeriver94 commented:
I hope I addressed all the comments, although I haven't got time to test it on a real cluster/openstack site yet. I will try to find the time during these days and let you know so you don't review untested code. I will mark the PR as draft for now

ezeriver94 commented:
Hi @mandre, after some struggle with e2e I believe it's ready for review again. Sorry for the commit spam, but I had to use the workflows you already have in place for deploying devstack as I don't have a local instance of it, and in the one I used for testing I don't have admin credentials.
There is one thing I wanted to hear your opinion about: SubnetRef and FloatingIP fields within spec.resource are both optional and you could use either one of them, but I was wondering if it makes sense to validate (only during resource creation) that if a FloatingIP is specified, the subnet should also be defined (even if it's not mandatory in gophercloud). My question relates to the fact that the IP must belong to a subnet in the network so it would be weird that you know which IP is valid without knowing if a proper subnet is part of the referenced network.

mandre commented on cmd/resource-generator/main.go:
```diff
@@ -40,6 +40,7 @@ type templateFields struct {
 	Year                   string
 	Name                   string
 	NameLower              string
+	IsNotNamed             bool
```
I'm not a big fan of new field's name, but I can't come up with a better name. Perhaps we could just make `getResourceName()` gracefully handle the case where the spec does not have a name instead?

mandre reviewed (commented):
This is really close, we should be able to merge soon. I've got a few more comments, the most important one being for the name of the network reference. I'm not sure if we should call it `networkRef`, as currently implemented, or `floatingNetworkRef` to be closer to neutron API.

Before we merge your patch, could you also squash the commits?

What happened to the `floatingip-import-error` test you had at some point? From what I recall, they were good.

ezeriver94 commented:
> What happened to the `floatingip-import-error` test you had at some point? From what I recall, they were good.

I believe at some moment during the e2e nightmare I decided it was not useful, but now I can't recall why. I'll create it again

>  I'm not sure if we should call it networkRef, as currently implemented, or floatingNetworkRef to be closer to neutron API.

About that, I don't have a clear answer, I'm fine with naming it as you prefer.

> I'm not a big fan of new field's name, but I can't come up with a better name. Perhaps we could just make getResourceName() gracefully handle the case where the spec does not have a name instead?

I can make getResourceName use reflection to detect if the ResourceSpec contains a name or not, I didn't do it this way before as it would be more disruptive, and I believe getResourceName should not be even defined for nameless resources, to make sure we don't use it during reconciliation (otherwise we would need to consider getResourceName to return empty string in multiple places, something that currently can't happen as all named resources will either have Resource.Name or Metadata.Name)
```go
func getResourceName(orcObject orcObjectPT) string {
	if orcObject.Spec.Resource.Name != nil {
		return string(*orcObject.Spec.Resource.Name)
	}
	return orcObject.Name
}
```

ezeriver94 commented:
> Before we merge your patch, could you also squash the commits?

done 😄

mdbooth commented on api/v1alpha1/floatingip_types.go:
```diff
@@ -0,0 +1,133 @@
+
+	// networkRef is a reference to the ORC Network which this resource is associated with.l
+	// +optional
+	NetworkRef KubernetesNameRef `json:"networkRef"`
```
Agree. It will make it more obvious what it is.

ezeriver94 commented on api/v1alpha1/floatingip_types.go:
```diff
@@ -0,0 +1,133 @@
+
+	// subnetRef references the subnet to which the floatingip is associated.
+	// +optional
+	SubnetRef *KubernetesNameRef `json:"subnetRef,omitempty"`
```
Just to be on the same page, @mdbooth @mandre, based on this other comment, do you still think I should rename to **FloatingNetworkRef**?

mandre replied:
I think yes. I'll give you a definitive answer later today after I've had time to think more about it.

mandre replied:
Indeed, after I it a bit more thoughts, I believe these fields should be called `FloatingNetworkRef` and `FloatingSubnetRef`.

Also, @mdbooth raised a good point yesterday. We should make `FloatingNetworkRef` and `FloatingSubnetRef` mutually exclusive, with exactly one being set, because a subnet always has a network, and letting users set a network at the same time as the subnet is only a potential source of error.

I _think_ the following validation enforces it:
```
// +kubebuilder:validation:XValidation:rule="has(self.floatingNetworkRef) != has(self.floatingSubnetRef)",message="Exactly one of 'floatingNetworkRef' or 'floatingSubnetRef' must be set"
```

mandre reviewed (commented):
If you could also improve the commit message, trying to stick to the [kubernetes guidelines](https://www.kubernetes.dev/docs/guide/pull-requests/#commit-message-guidelines) as much as possible.

ezeriver94 commented:
> If you could also improve the commit message, trying to stick to the [kubernetes guidelines](https://www.kubernetes.dev/docs/guide/pull-requests/#commit-message-guidelines) as much as possible.

sorry, I didn't notice the squashed commits remained. Will do

ezeriver94 commented:
@mandre I believe all comments were addressed (I hope I didn't miss anyone).
There is only one thing without a resolution:

> > I'm not a big fan of new field's name, but I can't come up with a better name. Perhaps we could just make getResourceName() gracefully handle the case where the spec does not have a name instead?
> 
> I can make getResourceName use reflection to detect if the ResourceSpec contains a name or not, I didn't do it this way before as it would be more disruptive, and I believe getResourceName should not be even defined for nameless resources, to make sure we don't use it during reconciliation (otherwise we would need to consider getResourceName to return empty string in multiple places, something that currently can't happen as all named resources will either have Resource.Name or Metadata.Name)
> 
> ```go
> func getResourceName(orcObject orcObjectPT) string {
> 	if orcObject.Spec.Resource.Name != nil {
> 		return string(*orcObject.Spec.Resource.Name)
> 	}
> 	return orcObject.Name
> }
> ```

Please let me know if you believe I should change something related to this template.

Related to 

> Also, @mdbooth raised a good point yesterday. We should make FloatingNetworkRef and FloatingSubnetRef mutually exclusive, with exactly one being set, because a subnet always has a network, and letting users set a network at the same time as the subnet is only a potential source of error.
> 
> I think the following validation enforces it:
> 
> // +kubebuilder:validation:XValidation:rule="has(self.floatingNetworkRef) != has(self.floatingSubnetRef)",message="Exactly one of 'floatingNetworkRef' or 'floatingSubnetRef' must be set"
> 

I added one more test called `floatingip-create-subnet-ref` as now the create-full did not cover subnetRef (being mutually exclusive with networkRef)

mandre commented:
> @mandre I believe all comments were addressed (I hope I didn't miss anyone). There is only one thing without a resolution:
> 
> > > I'm not a big fan of new field's name, but I can't come up with a better name. Perhaps we could just make getResourceName() gracefully handle the case where the spec does not have a name instead?
> > 
> > 
> > I can make getResourceName use reflection to detect if the ResourceSpec contains a name or not, I didn't do it this way before as it would be more disruptive, and I believe getResourceName should not be even defined for nameless resources, to make sure we don't use it during reconciliation (otherwise we would need to consider getResourceName to return empty string in multiple places, something that currently can't happen as all named resources will either have Resource.Name or Metadata.Name)
> > ```go
> > func getResourceName(orcObject orcObjectPT) string {
> > 	if orcObject.Spec.Resource.Name != nil {
> > 		return string(*orcObject.Spec.Resource.Name)
> > 	}
> > 	return orcObject.Name
> > }
> > ```
> 
> Please let me know if you believe I should change something related to this template.

Nah, it's alright. We don't expose it as an API so we can easily change it later if we don't like it.

> Related to
> 
> > Also, @mdbooth raised a good point yesterday. We should make FloatingNetworkRef and FloatingSubnetRef mutually exclusive, with exactly one being set, because a subnet always has a network, and letting users set a network at the same time as the subnet is only a potential source of error.
> > I think the following validation enforces it:
> > // +kubebuilder:validation:XValidation:rule="has(self.floatingNetworkRef) != has(self.floatingSubnetRef)",message="Exactly one of 'floatingNetworkRef' or 'floatingSubnetRef' must be set"
> 
> I added one more test called `floatingip-create-subnet-ref` as now the create-full did not cover subnetRef (being mutually exclusive with networkRef)

Good call. Let's get this merged :)
